File: tools/gen-manifest.py
#!/usr/bin/env python3
# ═══════════════════════════════════════════════════════════
#  SpellSpun Threads — manifest generator (Python)
#  Scans res/crochet/ and writes res/crochet/manifest.json so
#  the gallery loads with zero API calls (reliable on GitHub
#  Pages, offline, and any static server). Run after adding
#  photos:
#
#      python tools/gen-manifest.py
# ═══════════════════════════════════════════════════════════
import json
import re
import sys
from pathlib import Path

# Windows consoles default to cp1252; force UTF-8 so checkmark/warning glyphs print.
try:
    sys.stdout.reconfigure(encoding="utf-8")
except Exception:
    pass

ROOT = Path(__file__).resolve().parent.parent
CROCHET_DIR = ROOT / "res" / "crochet"
OUT = CROCHET_DIR / "manifest.json"

IMG_EXT = {
    "jpg", "jpeg", "png", "webp", "gif", "avif",
    "bmp", "heic", "heif", "svg", "tiff", "tif", "phn",
}
DIR_RE = re.compile(r"^(\d+)_(.*)$")


def lead_num(name: str):
    m = re.match(r"^\s*(\d+)", name)
    return int(m.group(1)) if m else float("inf")


def title_case(s: str) -> str:
    s = re.sub(r"[_-]+", " ", s)
    s = re.sub(r"\s+", " ", s).strip()
    return " ".join(w[0].upper() + w[1:].lower() for w in s.split() if w)


def main():
    if not CROCHET_DIR.is_dir():
        print(f"✗ Folder not found: {CROCHET_DIR}")
        sys.exit(1)

    collections = []
    for entry in sorted(CROCHET_DIR.iterdir()):
        if not entry.is_dir():
            continue
        m = DIR_RE.match(entry.name)
        if not m:
            print(f'  ⚠ skipping "{entry.name}" (does not match <number>_<title>)')
            continue
        order = int(m.group(1))
        title = title_case(m.group(2)) or entry.name
        files = [
            f.name for f in entry.iterdir()
            if f.is_file() and f.suffix.lower().lstrip(".") in IMG_EXT
        ]
        files.sort(key=lambda n: (lead_num(n), n))
        if not files:
            print(f'  ⚠ "{entry.name}" has no images — skipped')
            continue
        collections.append({"dir": entry.name, "order": order, "title": title, "images": files})

    collections.sort(key=lambda c: (c["order"], c["dir"]))
    OUT.write_text(json.dumps(collections, indent=2, ensure_ascii=False) + "\n", encoding="utf-8")

    photos = sum(len(c["images"]) for c in collections)
    print(f"✓ Wrote {OUT}")
    print(f"  {len(collections)} collection{'s' if len(collections) != 1 else ''}, "
          f"{photos} photo{'s' if photos != 1 else ''}")
    for c in collections:
        print(f"    {str(c['order']).rjust(2)}  {c['title'].ljust(18)} {len(c['images'])} photo(s)")
    if not collections:
        print('  (empty — add folders like res/crochet/1_star/ with photos inside)')


if __name__ == "__main__":
    main()

File: tools/gen-manifest.mjs
#!/usr/bin/env node
// ═══════════════════════════════════════════════════════════
//  SpellSpun Threads — manifest generator
//  Scans res/crochet/ and writes res/crochet/manifest.json
//  so the gallery loads with zero API calls (100% reliable on
//  GitHub Pages and offline). Run before pushing new photos:
//
//      node tools/gen-manifest.mjs
// ═══════════════════════════════════════════════════════════
import { readdir, writeFile } from 'node:fs/promises';
import { fileURLToPath } from 'node:url';
import { dirname, join } from 'node:path';

const __dirname = dirname(fileURLToPath(import.meta.url));
const ROOT = join(__dirname, '..');
const CROCHET_DIR = join(ROOT, 'res', 'crochet');
const OUT = join(CROCHET_DIR, 'manifest.json');

const IMG_EXT = new Set(['jpg','jpeg','png','webp','gif','avif','bmp','heic','heif','svg','tiff','tif','phn']);
const isImage = f => IMG_EXT.has((f.split('.').pop() || '').toLowerCase());
const leadNum = s => { const m = String(s).match(/^\s*(\d+)/); return m ? parseInt(m[1], 10) : Infinity; };
const titleCase = s => s.replace(/[_-]+/g, ' ').replace(/\s+/g, ' ').trim()
  .replace(/\w\S*/g, w => w.charAt(0).toUpperCase() + w.slice(1).toLowerCase());

async function scan() {
  let entries;
  try { entries = await readdir(CROCHET_DIR, { withFileTypes: true }); }
  catch { console.error(`✗ Folder not found: ${CROCHET_DIR}`); process.exit(1); }

  const collections = [];
  for (const ent of entries) {
    if (!ent.isDirectory()) continue;
    const m = ent.name.match(/^(\d+)_(.*)$/);
    if (!m) {
      console.warn(`  ⚠ skipping "${ent.name}" (does not match <number>_<title>)`);
      continue;
    }
    const order = parseInt(m[1], 10);
    const title = titleCase(m[2]) || ent.name;
    const sub = join(CROCHET_DIR, ent.name);
    const files = (await readdir(sub, { withFileTypes: true }))
      .filter(f => f.isFile() && isImage(f.name))
      .map(f => f.name)
      .sort((a, b) => (leadNum(a) - leadNum(b)) || a.localeCompare(b));

    if (!files.length) { console.warn(`  ⚠ "${ent.name}" has no images — skipped`); continue; }
    collections.push({ dir: ent.name, order, title, images: files });
  }

  collections.sort((a, b) => a.order - b.order || a.dir.localeCompare(b.dir));
  return collections;
}

const data = await scan();
await writeFile(OUT, JSON.stringify(data, null, 2) + '\n', 'utf8');

const photos = data.reduce((n, c) => n + c.images.length, 0);
console.log(`✓ Wrote ${OUT}`);
console.log(`  ${data.length} collection${data.length === 1 ? '' : 's'}, ${photos} photo${photos === 1 ? '' : 's'}`);
data.forEach(c => console.log(`    ${String(c.order).padStart(2,'0')}  ${c.title.padEnd(18)} ${c.images.length} photo(s)`));
if (!data.length) console.log('  (empty — add folders like res/crochet/1_star/ with photos inside)');
